Compile BPF code using vmlinux.h

wazir-ahmed opened the pull request:
If the host has kernel BTF information (`/sys/kernel/btf/vmlinux`),
then generate `vmlinux.h` using `bpftool` and compile the BPF code
using `vmlinux.h`.

This removes dependency on kernel headers to be present in the host
system if it uses a kernel > v5.4.

Fixes #803 

Signed-off-by: Wazir Ahmed <wazir@accuknox.com>

daemon1024 commented on KubeArmor/BPF/enforcer.bpf.c:
```diff
@@ -13,6 +13,10 @@ char LICENSE[] SEC("license") = "Dual BSD/GPL";
+/* In some of the distro, the generated
+ * vmlinux.h does not contain this value */
+#define PROC_PID_INIT_INO   0xEFFFFFFCU
+
```
Wouldn't this overwrite if present?
Perhaps should be something like:
```suggestion
#ifndef  PROC_PID_INIT_INO
/* In some of the distro, the generated
 * vmlinux.h does not contain this value */
#define PROC_PID_INIT_INO   0xEFFFFFFCU
#endif

```

wazir-ahmed replied:
In [<linux/proc_ns.h>](https://elixir.bootlin.com/linux/latest/source/include/linux/proc_ns.h#L46), `PROC_PID_INIT_INO` is defined as an `enum`. We cannot use `ifdef` or `ifndef` with enum constants.

daemon1024 replied:
Okay! Then the question is would this overwrite? the enum definition?

wazir-ahmed replied:
Yes. We cannot check for its existence, but we can overwrite enum constants with `#define`.

daemon1024 replied:
It should be fine, unless Some future version of the kernel updates the version 👍🏽

wazir-ahmed committed:
Compile BPF code using vmlinux.h

If the host has kernel BTF information (`/sys/kernel/btf/vmlinux`),
then generate `vmlinux.h` using `bpftool` and compile the BPF code
using `vmlinux.h`.

This removes dependency on kernel headers to be present in the host
system if it uses a kernel > v5.4.

Signed-off-by: Wazir Ahmed <wazir@accuknox.com>

daemon1024 committed:
Include version.h for non BTF supported compilation

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

daemon1024 committed:
use core read when using vmlinux.h

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

wazir-ahmed commented on KubeArmor/BPF/Makefile:
```diff
@@ -33,20 +33,27 @@ endif
 	@echo "Compiling eBPF bytecode: $(GREEN)$@$(NC) ..."
 	@echo "Using Compiler flags: $(CFlags)"
-	$(Q)$(CL) $(KF) -DMONITOR_HOST_AND_CONTAINER $(CFlags) -c $< -o -| llc -march=bpf -mcpu=probe -filetype=obj -o $@
+	$(Q)$(CL) $(KF) -DMONITOR_HOST_AND_CONTAINER $(CFlags) -Xclang -disable-llvm-passes -c $< -o - | opt -O2 -mtriple=bpf-pc-linux | llvm-dis | llc -march=bpf -mcpu=probe -filetype=obj -o $@
```
We need to use the same compilation command in `tests/Makefile` as well. Check the `syscall_test` target above.

daemon1024 replied:
We don't use any CORE reads so it should be fine imo. This is needed to process CORE IR.

achrefbensaad reviewed (approved):
LGTM, @wazir-ahmed  Can you please add the installation of bpftool to the setup script contribution/self-managed-k8s/setup.sh ?

wazir-ahmed committed:
Added linux-tools package in setup.sh

Signed-off-by: Wazir Ahmed <wazir@accuknox.com>

wazir-ahmed commented on contribution/self-managed-k8s/setup.sh:
```diff
@@ -21,7 +21,7 @@ export DEBIAN_FRONTEND=noninteractive
 
 # install dependencies and llvm--toolchain
-sudo apt-get -y install build-essential libelf-dev pkg-config net-tools linux-headers-generic
+sudo apt-get -y install build-essential libelf-dev pkg-config net-tools linux-headers-$(uname -r) linux-tools-$(uname -r)
```
`linux-*-generic` packages always points to the default kernel the OS comes with. If the user updates the kernel, then it still points to older kernel version.

wazir-ahmed replied:
@achrefbensaad Updated `setup.h`. The `linux-tools` package in Ubuntu has `bpftool`.
